pkg/apis/contrail/v1alpha1/contrailcni_types.go

```diff
@@ -0,0 +1,165 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ContrailCNI is the Schema for the contrailcnis API
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=contrailcnis,scope=Namespaced
+type ContrailCNI struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ContrailCNISpec   `json:"spec,omitempty"`
+	Status ContrailCNIStatus `json:"status,omitempty"`
+}
+
+// ContrailCNISpec defines the desired state of ContrailCNI
+type ContrailCNISpec struct {
+	CommonConfiguration  CNIPodConfiguration      `json:"commonConfiguration,omitempty"`
+	ServiceConfiguration ContrailCNIConfiguration `json:"serviceConfiguration"`
+}
+
+//ContrailCNIConfiguration is the Service Configuration for ContrailCNI
+// +k8s:openapi-gen=true
+type ContrailCNIConfiguration struct {
+	Containers      []*Container `json:"containers,omitempty"`
+	ControlInstance string       `json:"controlInstance,omitempty"`
+	CniMetaPlugin   string       `json:"cniMetaPlugin,omitempty"`
+	VrouterIP       string       `json:"vrouterIP,omitempty"`
+	VrouterPort     *int32       `json:"vrouterPort,omitempty"`
+	PollTimeout     *int32       `json:"pollTimeout,omitempty"`
+	PollRetries     *int32       `json:"pollRetries,omitempty"`
+	LogLevel        *int32       `json:"logLevel,omitempty"`
+}
+
+//CNIPodConfiguration is the Common Configuration for ContrailCNI
+// +k8s:openapi-gen=true
+type CNIPodConfiguration struct {
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/.
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty" protobuf:"bytes,7,rep,name=nodeSelector"`
+	// Host networking requested for this pod. Use the host's network namespace.
+	// If this option is set, the ports that will be used must be specified.
+	// Default to false.
+	// +k8s:conversion-gen=false
+	// +optional
+	HostNetwork *bool `json:"hostNetwork,omitempty" protobuf:"varint,11,opt,name=hostNetwork"`
+	// ImagePullSecrets is an optional list of references to secrets in the same namespace to use for pulling any of the images used by this PodSpec.
+	ImagePullSecrets []string `json:"imagePullSecrets,omitempty"`
+	// If specified, the pod's tolerations.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty" protobuf:"bytes,22,opt,name=tolerations"`
+}
+
+// ContrailCNIStatus defines the observed state of ContrailCNI
+type ContrailCNIStatus struct {
+	Active *bool `json:"active,omitempty"`
```

Comment:
I am not sure if its applicable here. But we are using generic status in our resources:
https://github.com/Juniper/contrail-operator/blob/a1010f2ef7e636b020e90b58b0927d9549de97f9/pkg/apis/contrail/v1alpha1/memcached_types.go#L15

We also have customised print columns:
https://github.com/Juniper/contrail-operator/blob/a1010f2ef7e636b020e90b58b0927d9549de97f9/pkg/apis/contrail/v1alpha1/memcached_types.go#L33